text/000-bundle-field-api.md

```diff
@@ -0,0 +1,159 @@
+
+# Summary
+[summary]: #summary
+
+This RFC proposes add `Field` as an extension to the `Bundle` construction API to specify
+precisely those parts of a `Bundle` that are actually intended to be functioning fields
+
+# Motivation
+[motivation]: #motivation
+
+The main goal of this addition is to further reduce the number of situations
+where a developer needs to write a `cloneType` method for custom `Bundle`s.
+It is further motivated by the difficulty in recognizing problematic constructions and
+crafting comprehensible error messages when they occur.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+It is proposed that an optional binding operator `Field` be added to the `Bundle` construction API.
+`Field` is similar to the binding operators `IO`, `Input`, `Output`, etc.
+It will allow with better semantics, which will coexist with the old semantics indefinitely. 
+
+Simple example:
+```scala
+class MyBundle extends Bundle {
+  val x = UInt(32.W)
+  val y = UInt(16.W)
+  val flag = Bool()
+}
+```
+would optionally become.
+```scala
+class MyBundle extends Bundle {
+  val x = Field(UInt(32.W))
+  val y = Field(UInt(16.W))
+  val flag = Field(Bool())
+}
+```
+
+Notable in this the following, `gen` and `constant` are not treated as fields of `MyFancierBundle`
+
+```scala
+class MyFancierBundle(val gen: Data) extends Bundle {
+  val x = Field(UInt(32.W))
+  val y = Field(UInt(16.W))
+  val flag = Field(Bool())
+  val constant = 7.U
+}
+```
+
+The new semantics are, tentatively:
+
+`Fields` are all or nothing, if you define any `Bundle` element as a `Field`, you must define all Bundle elements as Fields. The compiler will error out if you either have vals of type Data that aren't declared as a Field, or if you have Fields that weren't autodetected to be Bundle fields (which is a limitation of reflection-based detection).
+As it's not possible to have a Field(...) declaration in a constructor argument, constructor Data vals will NOT be treated as fields.
+`Field` will be a Bundle-specific API, and cannot be called outside a Bundle.
+This should eliminate all uses of cloneType in non-experimental Chisel constructs.
+
+In experimental, Record may still need cloneType, but Record is an advanced construct, and we may redefine it in the future to be more friendly.
+There was discussion about changing the name of this new API, for example as Struct (or something) instead of Bundle. I'm inclined to stick with Bundle, but there may be a discussion to be continued here.
+We will encourage the use of Field in new code, and there is no set timeline for when (if ever) old style code will generate deprecations, warnings, or errors.
+
+Suggestions are welcome!
+
+# Background
+[background]: #background
+
+The standard pattern for using Chisel Data objects as generator parameters in `Bundle`s is to pass 
+them as a constructor argument like so:
+```scala
+class MyBundle[T <: Data](gen: T) extends Bundle {
+  val foo = Input(gen)
```

Comment:
you need a chiselTypeOf() or .cloneType here

Reply:
You only need a `chiselTypeOf` if you're intending to pass bound hardware as `gen`, but the standard style is to use unbound types so this is correct.